[package] Use custom persistent_load in PackageImporter

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51489 [fx/package] make GraphModules packageable
* #51420 [fx] Make imports more selective in imported code.
* #51826 [package] make torch.package produce unified format
* **#51595 [package] Use custom persistent_load in PackageImporter**
* #51694 [package] more reliable method for determining standard library-ness

Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

suo committed:
[package] Use custom persistent_load in PackageImporter

Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

[ghstack-poisoned]

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

zdevito commented on torch/package/importer.py:
```diff
@@ -144,7 +153,30 @@ def load_pickle(self, package: str, resource: str, map_location=None) -> Any:
+            assert typename == 'storage', \
+                f"Unknown typename for persistent_load, expected 'storage' but got '{typename}'"
+            data_type, key, location, size = data
+            if key not in self._loaded_storages:
```
if you call `load_pickle` twice on the same resource you will get different copies of the model, but they will share tensor data. That seems inconsistent because it is neither 100% a copy nor 100% an alias. My intuition is that the tensors should be copies across different calls to load.

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

zdevito commented on torch/package/importer.py:
```diff
@@ -144,7 +145,38 @@ def load_pickle(self, package: str, resource: str, map_location=None) -> Any:
+        unpickler.persistent_load = persistent_load
+        result = unpickler.load()
+
+        torch._utils._validate_loaded_sparse_tensors()
```
I know that this was already in `_load` and this is just unsplitting code that will diverge, but this stateful weird function will need to be removed in our efforts to unify the format.

zdevito replied:
(it has a race condition if multiple python threads try to read independent files anyway)

suo replied:
will add a todo I guess?

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]

suo committed:
Update on "[package] Use custom persistent_load in PackageImporter"


Right now `PackageExporter` defines its own `persistent_id` but
`PackageImporter` uses the one defined in `torch.serialization`. I have
some downstream plans to customize this so this PR just splits it out.

Not to fear! I know this introduces some duplication and potential for
different behavior between `torch.save` and `torch.package`, but I have
plans to re-unify them soon.

Differential Revision: [D26211578](https://our.internmc.facebook.com/intern/diff/D26211578)

[ghstack-poisoned]
